mmdet/core/hook/wandblogger_hook.py

```diff
@@ -0,0 +1,462 @@
+                 log_checkpoint_metadata=False,
+                 num_eval_images=100,
+                 **kwargs):
+        super(WandbLogger, self).__init__(wandb_init_kwargs, interval,
+                                          **kwargs)
+
+        self.log_checkpoint = log_checkpoint
+        self.log_checkpoint_metadata = log_checkpoint_metadata
+        self.num_eval_images = num_eval_images
+        self.log_eval_metrics = True
+        self.best_score = 0
+        self.val_step = 0
+
+    @master_only
+    def before_run(self, runner):
+        super(WandbLogger, self).before_run(runner)
+        self.cfg = self.wandb.config
+        # Check if configuration is passed to wandb.
+        if len(dict(self.cfg)) == 0:
+            warnings.warn(
+                'To log mmdetection Config, '
+                'pass it to init_kwargs of WandbLogger.', UserWarning)
+
+        # Check if EvalHook and CheckpointHook are available.
+        for hook in runner.hooks:
+            if isinstance(hook, CheckpointHook):
+                self.ckpt_hook = hook
+            if isinstance(hook, (EvalHook, DistEvalHook)):
+                self.eval_hook = hook
+
+        # If CheckpointHook is not available turn off log_checkpoint.
+        try:
+            self._check_priority(self.ckpt_hook)
+        except AttributeError:
+            self.log_checkpoint = False
+            warnings.warn('To use log_checkpoint turn use '
+                          'CheckpointHook.', UserWarning)
+
+        # If EvalHook/DistEvalHook is not present set
+        # num_eval_images to zero.
+        try:
+            self.val_dataloader = self.eval_hook.dataloader
+            self.val_dataset = self.val_dataloader.dataset
+            self._check_priority(self.eval_hook)
+        except AttributeError:
+            self.num_eval_images = 0
+            self.log_checkpoint_metadata = False
+            self.log_eval_metrics = False
+            warnings.warn(
+                'To log num_eval_images turn validate '
+                'to True in train_detector.', UserWarning)
+
+        # If num_eval_images is greater than zero, create
+        # and log W&B table for validation data.
+        if self.num_eval_images > 0:
+            # Initialize data table
+            self._init_data_table()
+            # Add data to the table
+            self._add_ground_truth()
+            # Log ground truth data
+            self._log_data_table()
+
+        # Define a custom x-axes for validation metrics.
+        if self.log_eval_metrics:
+            self.wandb.define_metric('val/val_step')
+            self.wandb.define_metric('val/*', step_metric='val/val_step')
+
+    @master_only
+    def after_train_epoch(self, runner):
+        if self.log_eval_metrics:
+            if self.eval_hook.by_epoch:
+                if self.every_n_epochs(
+                        runner,
+                        self.eval_hook.interval) or self.is_last_epoch(runner):
```

Comment:
We can use `if self.eval_hook.by_epoch and self.eval_hook._should_evaluate(runner):`

referring to https://github.com/open-mmlab/mmcv/blob/68a2c0a1e2b0b87cd81b71a93eac5eb301cce5be/mmcv/runner/hooks/evaluation.py#L279-L312